mayastor/src/bdev/nexus/nexus_bdev.rs

```diff
@@ -1047,61 +1046,85 @@ pub async fn nexus_create(
+    nexus_list.push(Nexus::new(name, size, uuid, None));
+
+    // Obtain a reference to the newly created Nexus object.
+    let ni =
+        nexus_list
+            .iter_mut()
+            .find(|n| n.name == name)
+            .ok_or_else(|| Error::NexusNotFound {
+                name: String::from(name),
+            })?;
 
     for child in children {
-        if let Err(err) = ni.create_and_register(child).await {
-            error!("failed to create child {}: {}", child, err);
-            ni.destroy_children().await;
-            return Err(err).context(CreateChild {
-                name: ni.name.clone(),
+        if let Err(error) = ni.create_and_register(child).await {
+            error!(
+                "failed to create nexus {}: failed to create child {}: {}",
+                name, child, error
+            );
+            ni.close_children().await;
+            nexus_list.retain(|n| n.name != name);
+            return Err(Error::CreateChild {
+                source: error,
+                name: String::from(name),
             });
         }
     }
 
     match ni.open().await {
-        // we still have code that waits for children to come online
-        // this however only works for config files so we need to clean up
-        // if we get the below error
         Err(Error::NexusIncomplete {
             ..
         }) => {
-            info!("deleting nexus due to missing children");
-            for child in children {
-                if let Err(e) = bdev_destroy(child).await {
-                    error!("failed to destroy child during cleanup {}", e);
-                }
-            }
-
-            return Err(Error::NexusCreate {
+            // We still have code that waits for children to come online,
+            // although this currently only works for config files.
+            // We need to explicitly clean up child bdevs if we get this error.
+            error!("failed to open nexus {}: missing children", name);
+            destroy_child_bdevs(name, children).await;
+            nexus_list.retain(|n| n.name != name);
+            Err(Error::NexusCreate {
                 name: String::from(name),
-            });
+            })
         }
 
-        Err(e) => {
-            error!("failed to open nexus {}: {}", ni.name, e);
-            ni.destroy_children().await;
-            return Err(e);
+        Err(error) => {
+            error!("failed to open nexus {}: {}", name, error);
+            ni.close_children().await;
+            nexus_list.retain(|n| n.name != name);
+            Err(error)
         }
 
-        Ok(_) => nexus_list.push(ni),
+        Ok(_) => Ok(()),
+    }
+}
+
+/// Destroy list of child bdevs
+async fn destroy_child_bdevs(name: &str, list: &[String]) {
+    let futures = list.iter().map(String::as_str).map(bdev_destroy);
+    let results = join_all(futures).await;
+    if results.iter().any(|c| c.is_err()) {
```

Comment:
This will finish after any failure, it seems we'd want to ensure all destruction commands are run and attempted to be completed?

Reply:
Ouch! Yes you are right. I hade mistakenly believed that with join_all(), all subfutures are driven to completion.

Reply:
Hmm ... According to the documentation, the *old* behaviour was that futures are cancelled immediately on first error. However, as of version 0.3.0, no mention of this is made in the docs any more - suggesting now that the current behaviour is exactly what we want.